crates/circuit/src/parameter_expression.rs

```diff
@@ -0,0 +1,507 @@
+                    } else {
+                        c.into_py_any(py)
+                    }
+                }
+            },
+            None => Err(pyo3::exceptions::PyRuntimeError::new_err(
+                "Expression has some undefined symbols.",
+            )),
+        }
+    }
+
+    /// clone expression
+    pub fn copy(&self) -> Self {
+        Self {
+            expr: self.expr.clone(),
+        }
+    }
+    /// return conjugate of expression
+    pub fn conjugate(&self) -> Self {
+        Self {
+            expr: self.expr.conjugate(),
+        }
+    }
+    /// return derivative of this expression for param
+    pub fn derivative(&self, param: &Self) -> PyResult<Self> {
+        match self.expr.derivative(&param.expr) {
+            Ok(expr) => Ok(Self { expr }),
+            Err(e) => Err(pyo3::exceptions::PyRuntimeError::new_err(e)),
+        }
+    }
+
+    /// expand expression
+    pub fn expand(&self) -> Self {
+        Self {
+            expr: self.expr.expand(),
+        }
+    }
+
+    /// get hashset of all the symbols used in this expression
+    pub fn symbols(&self) -> HashSet<String> {
+        self.expr.symbols()
+    }
+
+    /// return all values in this equation
+    pub fn values(&self, py: Python) -> PyResult<Vec<PyObject>> {
+        self.expr
+            .values()
+            .iter()
+            .map(|val| match val {
+                symbol_expr::Value::Real(r) => r.into_py_any(py),
+                symbol_expr::Value::Int(i) => i.into_py_any(py),
+                symbol_expr::Value::Complex(c) => c.into_py_any(py),
+            })
+            .collect()
+    }
+
+    /// return expression as a string
+    #[getter]
+    pub fn name(&self) -> String {
+        self.__str__()
+    }
+
+    /// bind values to symbols given by input hashmap
+    #[pyo3(name = "bind")]
+    pub fn py_bind(&self, map: HashMap<String, Bound<PyAny>>) -> PyResult<Self> {
+        let map: HashMap<String, symbol_expr::Value> = map
+            .into_iter()
+            .filter_map(|(key, val)| {
+                if let Ok(i) = val.extract::<i64>() {
+                    Some((key, symbol_expr::Value::from(i)))
+                } else if let Ok(r) = val.extract::<f64>() {
+                    Some((key, symbol_expr::Value::from(r)))
+                } else if let Ok(c) = val.extract::<Complex64>() {
+                    Some((key, symbol_expr::Value::from(c)))
+                } else {
+                    // if unsupported data type, insert nothing
+                    None
+                }
+            })
+            .collect();
+        let bound = self.expr.bind(&map);
+        match bound.eval(true) {
+            Some(v) => match &v {
+                symbol_expr::Value::Real(r) => {
+                    if r.is_infinite() {
+                        Err(pyo3::exceptions::PyZeroDivisionError::new_err(
+                            "zero division occurs while binding parameter",
```

Comment:
Technically, this does not need to come from a zero division while binding: the bound value can already be `inf`, e.g. due to some prior error in the users code, before dealing with parameters. For example
```python
x = Parameter("x")
x.bind({x: np.inf})
```
triggers this. Maybe we can reword this (and the one below too).

```suggestion
                            "attempted to bind infinite value to parameter",
```